mux/src/tmux_commands.rs

```diff
@@ -59,148 +84,535 @@ impl TmuxDomainState {
+
+        let _ = mux.add_pane(&pane);
+
+        return Ok(pane);
+    }
+
+    pub fn fix_attached_pane_id(
+        &self,
+        window_id: TmuxWindowId,
+        old_id: TmuxPaneId,
+        new_id: TmuxPaneId,
+    ) -> anyhow::Result<()> {
+        let mut pane_map = self.remote_panes.lock();
+        match pane_map.remove(&old_id) {
+            Some(ref_pane) => {
+                ref_pane.lock().pane_id = new_id;
+                pane_map.insert(new_id, ref_pane);
+                let _ = self.add_attached_pane(window_id, new_id);
+            }
+            None => {}
+        }
+
+        return Ok(());
+    }
+
     fn sync_pane_state(&self, panes: &[PaneItem]) -> anyhow::Result<()> {
-        // TODO:
-        // 1) iter over current session panes
-        // 2) create pane if not exist
-        // 3) fetch scroll buffer if new created
-        // 4) update pane state if exist
         let current_session = self.tmux_session.lock().unwrap_or(0);
+        let mux = Mux::get();
+
         for pane in panes.iter() {
-            if pane.session_id != current_session || self.check_pane_attached(&pane) {
+            if pane.session_id != current_session
+                || !self.check_pane_attached(pane.window_id, pane.pane_id)
+            {
                 continue;
             }
 
-            let local_pane_id = alloc_pane_id();
-            let (output_read, output_write) = filedescriptor::socketpair()?;
-            let active_lock = Arc::new((Mutex::new(false), Condvar::new()));
-
-            let ref_pane = Arc::new(Mutex::new(TmuxRemotePane {
-                local_pane_id,
-                output_write,
-                active_lock: active_lock.clone(),
-                session_id: pane.session_id,
-                window_id: pane.window_id,
-                pane_id: pane.pane_id,
-                cursor_x: pane.cursor_x,
-                cursor_y: pane.cursor_y,
-                pane_width: pane.pane_width,
-                pane_height: pane.pane_height,
-                pane_left: pane.pane_left,
-                pane_top: pane.pane_top,
-            }));
+            // We now have the cursor information, fix the cursor position
+            let pane_map = self.remote_panes.lock();
+            let local_pane = match pane_map.get(&pane.pane_id) {
+                Some(p) => {
+                    let local_pane_id = p.lock().local_pane_id;
+                    mux.get_pane(local_pane_id)
+                }
+                None => None,
+            };
 
-            {
-                let mut pane_map = self.remote_panes.lock();
-                pane_map.insert(pane.pane_id, ref_pane.clone());
+            if let Some(local_pane) = local_pane {
+                local_pane.set_cursor_position(pane.cursor_x as usize, pane.cursor_y as usize);
+                if pane.pane_active {
+                    let gui_tabs = self.gui_tabs.lock();
+                    let local_tab = gui_tabs.get(&pane.window_id).unwrap();
```

Comment:
bare unwrap here; let's convert to an error that we can report without crashing:

```suggestion
                    let Some(local_tab) = gui_tabs.get(&pane.window_id) else {
                       anyhow::bail!("invalid tmux window id {}", pane.window_id);
                    }
```